stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -93,6 +107,30 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
         args_adjoints_ = Eigen::VectorXd::Zero(num_vars_shared_terms_);
       }
 
+      // Obtain reference to a local copy of all shared arguments that do
+      // not point
+      //   back to main autodiff stack
+
+      if (!local_args_tuple_scope_.args_tuple_holder_) {
```

Comment:
The conditional checks if the local copy of the shared arguments has already been made for the particular reducer *instance*. The TBB may use more than one of these workers then there are thread, but the number of workers should be at the order of how many threads we have. Note that the TBB schedules work chunks by using the reducer instances; these work chunks (or tasks) are handed to the scheduler which executes them in whatever thread. Since reducers are used multiple times for multiple ranges it can happen that one reducer gets to run in different threads.

> How does it ensure that there is a different copy for each thread? 

We do not need to check for threads here at all anymore, since everything is referred to from the reduce instances (but only one instance of the reducers can ever be active in a given thread, of course).

The `mutex` in `scoped_chainablestack` will prevent that for a particular instance the `execute` is called simultaneously from different threads. Whenever that happens the program will terminate as it really must never happen.

Reply:
Great, makes sense thanks